### PR TITLE
Add support for `reinforce++` and `reinforce++-baseline`

### LINKED ISSUE #16: [feat] Supporting more RL algorithms.
Thanks to your astonishing RL framework ❤️. I was wondering if you have any plan to support more RL algorithms, such as reinforce++.

### PR DESCRIPTION
**Pull request by Chen-GX**
## Summary
This PR introduces support for the `reinforce++` and `reinforce++-baseline` algorithms into the Slime framework. It aims to evaluate their performance in comparison to the existing `grpo` implementation.

Closes #16 

## Key Changes

- Implemented the core logic for `reinforce++` and its baseline variant in `ppot_utils.py`.
- Implemented the `distributed_masked_whiten` for normalization in `slime/utils/distributed_utils.py`.
- Added new experiment scripts: `scripts/run-qwen3-4B-rf.sh` and `scripts/run-qwen3-4B-rf-baseline.sh`.

## Experimental Results & Observations

I conducted experiments comparing the three algorithms, with results logged to wandb.

As shown in the charts below, `reinforce++-baseline` and `grpo` exhibit similar performance curves. However, the `reinforce++` implementation shows a distinct behavior (e.g., its reward curve appears more volatile/lower).

I have double-checked my implementation, and it seems to align with the algorithm's principles. I suspect the difference might stem from hyperparameter sensitivity or a subtle implementation detail.

I would appreciate it if you could pay special attention to the logic in `ppot_utils.py` and `slime/utils/distributed_utils.py`. I'm ready to make corrections if any mistakes are found.

BTW, this PR focuses on `reinforce++` and `reinforce++-baseline`. I'm happy to contribute implementations for other algorithms, such as RLOO, in a future PR if there is interest.

image
image
image
image

### CODE REVIEW

File: slime/backends/megatron_utils/data.py [hunk 1]
```diff
@@ -280,6 +280,15 @@ def process_rollout_data(rollout_id, args, data_buffer):
...
         data["rewards"] = rewards
 
+    elif args.advantage_estimator == "reinforce_plus_plus_baseline" and args.rewards_normalization:
+        # group baseline subtraction for RF++-baseline
+        rewards = torch.tensor([r for r in rewards], dtype=torch.float)
+        rewards = rewards.reshape(-1, args.n_samples_per_prompt)
+        mean = rewards.mean(dim=-1, keepdim=True)
+        rewards = rewards - mean
+        rewards = rewards.flatten().tolist()
+        data["rewards"] = rewards
```
Comment (zhuzilin) [inline slime/backends/megatron_utils/data.py]: here we can merge the code with the abort normalization, e.g.

```python
if args.rewards_normalization:
    ...
    if args.advantage_estimator == "grpo" and args.grpo_std_normalization:
        std = rewards.std(dim=-1, keepdim=True)
        rewards = rewards / (std + 1e-6)
    ...
```

File: slime/backends/megatron_utils/loss.py [hunk 2]
```diff
@@ -147,6 +149,27 @@ def compute_advantages_and_returns(args):
...
+
+    elif args.advantage_estimator == "reinforce_plus_plus_baseline":
+        advantages = get_reinforce_plus_plus_baseline_advantages(
+            rewards=rewards,
+            kl=kl,
+            loss_masks=loss_masks,
+            response_lengths=response_lengths,
+            kl_coef=args.kl_coef,
+        )
+        returns = advantages
```
Comment (zhuzilin) [inline slime/backends/megatron_utils/loss.py]: could you check if the result is the same with CP > 1. It may be a little tricky to implement the context parallel version for those 2 algos.

File: slime/utils/ppo_utils.py [hunk 3]
```diff
@@ -119,3 +121,111 @@ def get_grpo_returns(
...
+    # Broadcast to get unwhitened advantages
+    unwhitened_advantages = [
+        torch.ones_like(kl_tensor) * reward_val - kl_coef * kl_tensor
+        for kl_tensor, reward_val in zip(kl, rewards)
+    ]
+    # Concatenate tensors for a global operation
+    if loss_masks is None:
+        loss_masks = [
+            torch.ones_like(adv) for adv in unwhitened_advantages
+        ]
```
Comment (zhuzilin) [inline slime/utils/ppo_utils.py]: I've removed the case where `loss_masks is None`, we don't need this anymore.

File: slime/utils/ppo_utils.py [hunk 4]
```diff
@@ -119,3 +121,111 @@ def get_grpo_returns(
...
+
+    all_advs = torch.cat(unwhitened_advantages)
+    all_masks = torch.cat(loss_masks)
+    
+    whitened_advs_flat = distributed_masked_whiten(all_advs, all_masks, shift_mean=True)
```
Comment (zhuzilin) [inline slime/utils/ppo_utils.py]: Can we move the `distributed_masked_whiten` to the `NotImplementedError` in `if args.normalize_advantages:`?

File: slime/backends/megatron_utils/data.py (hunk 1 above)
Comment (Chen-GX) [reply slime/backends/megatron_utils/data.py]: Thanks for the great suggestion! I've refactored the reward normalization logic to merge the common parts for grpo and reinforce_plus_plus_baseline, eliminating the duplicated code. This should be much cleaner and easier to maintain now.

File: slime/utils/ppo_utils.py (hunk 3 above)
Comment (Chen-GX) [reply slime/utils/ppo_utils.py]: Thanks for your reminder. I've removed it.

File: slime/backends/megatron_utils/loss.py (hunk 2 above)
Comment (Chen-GX) [reply slime/backends/megatron_utils/loss.py]: I've added the support for comtext parallel, and I'll test it before pr.

File: slime/utils/ppo_utils.py (hunk 4 above)
Comment (Chen-GX) [reply slime/utils/ppo_utils.py]: I've moved the distributed_masked_whiten logic to the if args.normalize_advantages: block in compute_advantages_and_returns. Accordingly, the helper functions now only return the unwhitened values. This makes the control flow much cleaner.
